src/compiler/transformers/declarations.ts

```diff
@@ -254,55 +264,117 @@ export function transformDeclarations(context: TransformationContext) {
+                return {
+                    isolatedDeclarations,
+                    useLocalInferenceTypePrint: true,
+                    resolver,
+                    localInferenceResolver,
+                    symbolTracker: undefined,
+                    host: undefined,
+                    ensureNoInitializer: createEnsureNoInitializer(emptySymbolTracker),
+                } as const;
+            }
+            else {
+                const host = context.getEmitHost();
+                const resolver: EmitResolver = context.getEmitResolver();
+                const symbolTracker = createSymbolTracker(resolver, host);
+                return {
+                    isolatedDeclarations,
+                    useLocalInferenceTypePrint: false,
+                    resolver,
+                    localInferenceResolver,
+                    symbolTracker,
+                    host,
+                    ensureNoInitializer: createEnsureNoInitializer(symbolTracker),
+                } as const;
+            }
+        }
+        else {
+            Debug.assert(context.kind === TransformationContextKind.FullContext);
+            const host = context.getEmitHost();
+            const resolver = context.getEmitResolver();
+            const symbolTracker: SymbolTracker = createSymbolTracker(resolver, host);
+            return {
+                isolatedDeclarations: false,
+                useLocalInferenceTypePrint: false,
+                localInferenceResolver: undefined,
+                resolver,
+                symbolTracker,
+                host,
+                ensureNoInitializer: createEnsureNoInitializer(symbolTracker),
+            } as const;
+        }
+
+        function createEnsureNoInitializer(symbolTracker: SymbolTracker) {
+            return function ensureNoInitializer(node: CanHaveLiteralInitializer) {
+                if (shouldPrintWithInitializer(node)) {
+                    return resolver.createLiteralConstValue(getParseTreeNode(node) as CanHaveLiteralInitializer, symbolTracker); // TODO: Make safe
+                }
+                return undefined;
+            };
+        }
+    }
+    function recordTypeReferenceDirectivesIfNecessary(typeReferenceDirectives: readonly [specifier: string, mode: ResolutionMode][] | undefined, requestingNode: Node | undefined): void {
         if (!typeReferenceDirectives) {
             return;
         }
-        necessaryTypeReferences = necessaryTypeReferences || new Set();
+        necessaryTypeReferences = necessaryTypeReferences || new Map();
         for (const ref of typeReferenceDirectives) {
-            necessaryTypeReferences.add(ref);
+            necessaryTypeReferences.set(ref, requestingNode);
         }
     }
 
     function trackReferencedAmbientModule(node: ModuleDeclaration, symbol: Symbol) {
+        // We forbid references in isolated declarations no need to report any errors on them
+        if (isolatedDeclarations) return;

```

Comment:
I’m not quite following this. This function doesn't just get called for literal triple slash references in the input file; it also gets called when you write an _import_ that resolves to an ambient module. And returning early here changes the emit in a potentially dangerous way, without an accompanying `isolatedDeclarations` error. For example, if I have `@types/node` installed and a source file like

```ts
export { readFile } from "fs";
```

This is legal according to the PR as it stands, but the emit under `--isolatedModules` loses the triple-slash reference that makes `"fs"` resolvable:

```diff
- /// <reference types="node" />
export { readFile } from "fs";
```

Based on the analog to `isolatedModules`, my expectation for how this flag would work was that

1. Enabling the flag wouldn’t change emit with `tsc --declaration`
2. Enabling the flag would leverage whole-program info to issue diagnostics in cases like the above, where information from other files (or too-complex type inference) was necessary to produce the correct emit.

Reply:
You are right, I will change the code to make TS  emit equivalent.

Just to understand the implications of not emitting these reference directives in `transpileDeclaration`: 

Without these ts will not be able to resolve import of modules that are declared in the declarations of other modules. Such as `node` having a `declare mdoule 'fs' { ... }`. (corect e on the terminology here 😅)

For node this is not actually that big an issue from what I can tell since `getAutomaticTypeDirectiveNames` will automatically add all packages in `@types`, and node types are defined there.

For types that come from regular imports that resolve to a package in `node_modules` that contains the appropriate declarations, this again is not an issue.

The only way the absence of these directives becomes a problem is if the types for a module are not in the usual resolution path, and if the module is not in the `typeRoots`.

Reply:
I’ve skimmed through most of the remaining changes to declarations.ts, and there are _many_ cases of

```ts
if (isolatedDeclarations) {
  // return something different
}
```

Were these all intended to be `if (context.kind === TransformationContextKind.Isolated)`?

Reply:
Also, just to be clear, I don’t think that matching emit here is sufficient; I do think `isolatedDeclarations` needs to issue an error. I used `@types/node` as a simple example, and you’re right that in practice it’s very unlikely for the end user not to already have `@types/node` loaded due to other references. But that’s totally circumstantial. We emit those triple slash references for a reason; we can’t just say they’re universally unnecessary and silently elide them.

Reply:
> Also, just to be clear, I don’t think that matching emit here is sufficient; I do think `isolatedDeclarations` needs to issue an error. I used `@types/node` as a simple example, and you’re right that in practice it’s very unlikely for the end user not to already have `@types/node` loaded due to other references. But that’s totally circumstantial. We emit those triple slash references for a reason; we can’t just say they’re universally unnecessary and silently elide them.

We can definitely error on it, the problem is I don't know what the workaround would be. 

We struggled previously with reference directives. TS will not just add them, but will also remove them if they are not needed for the declaration file. This is the kind of analysis we can't do in isolated declaration emit. We discussed and decided with @DanielRosenwasser that banning all reference directives in isolated declaration mode would be the way to go. 

But due to this ban, it means that a user would be stuck. You can't `export * from 'fs'` without a reference directive, but adding it would be an error because we don't know if we need to keep it in the declaration or not. 

> I’ve skimmed through most of the remaining changes to declarations.ts, and there are _many_ cases of
> 
> ```ts
> if (isolatedDeclarations) {
>   // return something different
> }
> ```
> 
> Were these all intended to be `if (context.kind === TransformationContextKind.Isolated)`?

I will check them again tomorrow, but the `if (isolatedDeclarations) ` should be in cases where the `if` triggers an error, or if an error would have been triggered already. In such cases declaration emit should not happen anyway.

Reply:
> TS will not just add them, but will also remove them if they are not needed for the declaration file.

I’m not sure I was aware of this. Sorry to jump into a conversation you’ve already had, but did we consider changing normal declaration emit to preserve all manually written reference directives?

Reply:
@andrewbranch This example shows the removal of the reference directive.

```ts
//index.ts
/// <reference types="node" />
export function x() { }
//index.d.ts
export declare function x(): void;
```

I am not sure about keeping them in either. Wouldn't that keep implementation details in the declaration file? I am fine with always keeping them. I don't think people are adding these anymore for other reasons.

Reply:
I think it’s no different from keeping `import "./foo"` statements in declaration emit. The user wrote it on purpose and it changes the shape of their program; it feels odd to me that we would elide it.

Reply:
I will change it to always preserve reference directives with `--isolatedDeclarations`